Fix bug with DB-Forge composite foreign keys

monkenWu opened the pull request:
**Description**
Fixes #4310 
Hi, I edited the `addForeignKey` method and `_processForeignKeys` method of `CodeIgniter\Database\Forge` so that they can define composite foreign keys in an array.

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPdocs
- [x] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

samsonasik commented:
Please run cs fixer

```bash
vendor/bin/php-cs-fixer fix system/Database/Forge.php
vendor/bin/php-cs-fixer fix tests/system/Database/Live/ForgeTest.php
```

monkenWu commented:
@samsonasik It's done.

samsonasik commented:
Please run rector as well:

```bash
vendor/bin/rector
```

monkenWu commented:
@samsonasik 
Hi, I fixed the problem of `_processForeignKeys` method when using `SQLSRV` Driver, modified the test case, and ran `vendor/bin/rector process` .

paulbalandan commented on system/Database/Forge.php:
```diff
@@ -403,27 +403,40 @@ public function addField($field)
-            throw new DatabaseException(lang('Database.fieldNotExists', [$fieldName]));
+        $errorNames = [];
+
+        if (is_array($fieldName)) {
+            foreach ($fieldName as $name) {
+                if (! isset($this->fields[$name])) {
+                    $errorNames[] = $name;
+                }
+            }
+        } elseif (! isset($this->fields[$fieldName])) {
+            $errorNames[] = $fieldName;
         }
```
You can just simply force this to array then iterate over. No need for if else condition.
```php
$fieldName = (array) $fieldName;

foreach ($fieldName as $name) {
    if (! isset($this->fields[$name])) {
        $errorNames[] = $name;
    }
}
```

paulbalandan commented on system/Database/Forge.php:
```diff
@@ -403,27 +403,40 @@ public function addField($field)
-            'field'    => $tableField,
-            'onDelete' => strtoupper($onDelete),
-            'onUpdate' => strtoupper($onUpdate),
+        if (! empty($errorNames)) {
```
```suggestion
        if ($errorNames !== []) {
```

paulbalandan commented on system/Database/Forge.php:
```diff
@@ -1208,11 +1221,25 @@ protected function _processForeignKeys(string $table): string
-            foreach ($this->foreignKeys as $field => $fkey) {
-                $nameIndex = $table . '_' . $field . '_foreign';
+            foreach ($this->foreignKeys as $fkey) {
+                if (is_array($fkey['field'])) {
+                    $nameIndex = $table . '_' . implode('_', $fkey['field']) . '_foreign';
+                } else {
+                    $nameIndex = $table . '_' . $fkey['field'] . '_foreign';
+                }
+                if (is_array($fkey['field'])) {
+                    $foreignKeyFiled = implode(', ', $this->db->escapeIdentifiers($fkey['field']));
+                } else {
+                    $foreignKeyFiled = $this->db->escapeIdentifiers($fkey['field']);
+                }
```
Since `field` is now an array, you can dispense the if else.

paulbalandan commented on system/Database/Forge.php:
```diff
@@ -1208,11 +1221,25 @@ protected function _processForeignKeys(string $table): string
+                } else {
+                    $foreignKeyFiled = $this->db->escapeIdentifiers($fkey['field']);
+                }
+                if (is_array($fkey['referenceField'])) {
+                    $referenceField = implode(', ', $this->db->escapeIdentifiers($fkey['referenceField']));
+                } else {
+                    $referenceField = $this->db->escapeIdentifiers($fkey['referenceField']);
+                }
```
You can possibly force convert to array to dispense the if else.

paulbalandan commented on system/Database/Forge.php:
```diff
@@ -1208,11 +1221,25 @@ protected function _processForeignKeys(string $table): string
+                }
 
                 $sql .= ",\n\tCONSTRAINT " . $this->db->escapeIdentifiers($nameIndex)
-                    . ' FOREIGN KEY(' . $this->db->escapeIdentifiers($field) . ') REFERENCES ' . $this->db->escapeIdentifiers($this->db->DBPrefix . $fkey['table']) . ' (' . $this->db->escapeIdentifiers($fkey['field']) . ')';
+                    . ' FOREIGN KEY(' . $foreignKeyFiled . ') REFERENCES ' . $this->db->escapeIdentifiers($this->db->DBPrefix . $fkey['referenceTable']) . ' (' . $referenceField . ')';
```
Can you use `sprintf` here to easily see the format of the SQL you're creating?

paulbalandan commented on system/Database/SQLSRV/Forge.php:
```diff
@@ -233,46 +233,6 @@ protected function _processColumn(array $field): string
-     *
-     * @return string
-     */
-    protected function _processForeignKeys(string $table): string
```
You cannot simply delete a protected method as userland code might be depending on this. If this is no longer needed, then deprecate the method.

paulbalandan reviewed (commented):
Irrespective of the position whether the change in signature will be accepted or not, I have left some comments. And please change the base branch to `codeigniter4:4.2` instead of `develop`.

samsonasik commented on system/Database/Forge.php:
```diff
@@ -1208,11 +1221,25 @@ protected function _processForeignKeys(string $table): string
+                    $nameIndex = $table . '_' . $fkey['field'] . '_foreign';
+                }
+                if (is_array($fkey['field'])) {
+                    $foreignKeyFiled = implode(', ', $this->db->escapeIdentifiers($fkey['field']));
```
I think "Filed" should be "Filled" with double `l`

monkenWu commented:
@paulbalandan  @samsonasik

Thanks for the review.
I corrected some spelling errors, improved code quality, and rolledback the `_processIndexes` method of the `SQLSRV` class.

paulbalandan commented:
@monkenWu it seems your PR branch's tail is still based on `develop`. Please rebase this against the current `4.2` branch so that changes only from your PR will appear here. If anything went wrong,, you can just recreate your PR against `4.2`

monkenWu commented:
Change the branch base to 4.2.
